src/translation-api.js

```diff
@@ -0,0 +1,120 @@
+import { and, eq } from 'drizzle-orm'
+import { kCreateWithDocId, kSelect } from './datatype/index.js'
+import { hashObject } from './utils.js'
+
+export const ktranslatedLanguageCodeToSchemaNames = Symbol(
+  'translatedLanguageCodeToSchemaNames'
+)
+export default class TranslationApi {
+  /** @type {Map<
+   * import('@mapeo/schema').TranslationValue['languageCode'],
+   * Set<import('@mapeo/schema/dist/types.js').SchemaName>>} */
+  #translatedLanguageCodeToSchemaNames = new Map()
+  #dataType
+  #table
+
+  /**
+   * @param {Object} opts
+   * @param {import('./datatype/index.js').DataType<
+   *   import('./datastore/index.js').DataStore<'config'>,
+   *   typeof import('./schema/project.js').translationTable,
+   *   'translation',
+   *   import('@mapeo/schema').Translation,
+   *   import('@mapeo/schema').TranslationValue
+   * >}  opts.dataType
+   * @param {typeof import('./schema/project.js').translationTable} opts.table
+   */
+  constructor({ dataType, table }) {
+    this.#dataType = dataType
+    this.#table = table
+    this.#dataType
+      .getMany()
+      .then((docs) => docs.map((doc) => this.index(doc)))
+      .catch((err) => {
+        throw new Error(`error loading Translation cache: ${err}`)
+      })
+  }
+
+  /**
+   * @param {import('@mapeo/schema').TranslationValue} value
+   */
+  async put(value) {
+    /* eslint-disable no-unused-vars */
+    const { message, ...identifiers } = value
+    const docId = hashObject(identifiers)
+    try {
+      const doc = await this.#dataType.getByDocId(docId)
+      return await this.#dataType.update(doc.versionId, value)
+    } catch (e) {
+      // @ts-ignore how can this be improved (we can maybe set "useUnknownInCatchVariables": false)
+      if (e.message !== 'Not found')
+        throw new Error(`Error on translation ${e}`)
+      // TODO: throw if the error is different from 'Not found'
+      return await this.#dataType[kCreateWithDocId](docId, value)
+    }
+  }
+
+  /**
+   * @param {import('type-fest').SetOptional<
+   * Omit<import('@mapeo/schema').TranslationValue,'schemaName' | 'message'>,
+   * 'fieldRef' | 'regionCode'>} value
+   */
```

Comment:
Nitpick: This currently returns `any`. Could we add a return type?

```suggestion
   * @returns {Promise<import('@mapeo/schema').Translation[]>}
   */
```